pennylane/devices/qubit/simulate.py

```diff
@@ -44,6 +45,34 @@ def measure_state_diagonalizing_gates(
     return measurementprocess.process_state(qml.math.flatten(state), wires)
 
 
+def measure_hamiltonian_expval(
+    measurementprocess: MeasurementProcess, state: TensorLike
+) -> TensorLike:
+    """Measure the expectation value of the state when the measured observable is a ``Hamiltonian`` or
+    ``SparseHamiltonian``.
+
+    Args:
+        measurementprocess (MeasurementProcess): measurement process to apply to the state
+        state (TensorLike): the state to measure
+
+    Returns:
+        TensorLike: the result of the measurement
+    """
+    # Add case for backprop later
+    total_wires = len(state.shape)
+    Hmat = measurementprocess.obs.sparse_matrix(wire_order=list(range(total_wires)))
+    _state = qml.math.toarray(state).flatten()
+
+    # Find the expectation value using the <\psi|H|\psi> matrix contraction
+    bra = csr_matrix(qml.math.conj(_state))
+    ket = csr_matrix(_state[..., None])
+    new_ket = csr_matrix.dot(Hmat, ket)
+    res = csr_matrix.dot(bra, new_ket).toarray()[0]
```

Comment:
Some avenues to explore while we are re-examining this code.

1. Do we need to convert the state to `csr_matrix` first? Or can we just use dot a scipy csr matrix with a numpy array? Which one is more efficient?
2. If we do need to convert the state to a `csr_matrix` first, is it better to conjugate it in numpy form or csr form? `csr_matrix(state).conj()`

Reply:
After a very short traversal through stack overflow, [this answer] gives a good idea of the performance difference between `numpy.dot` and `csr_matrix.dot`. The latter seems to be faster as a matrix gets more sparse. In most cases, I think that a `Hamiltonian`, `Sum`, or `SparseHamiltonian` will be reasonably sparse that we might have a speedup using `csr_matrix.dot`.

That being said, the testing on the stack overflow answer uses VERY large matrices (5000x5000), so it's also possible that for smaller matrices numpy is faster, or that the difference between `np.dot` and `csr_matrix.dot` is negligible. I think some benchmarking would be necessary, but I'm happy leaving the code as is for now. The bottom line is that for less than ~10 wires, the difference in performance using either method would likely be insignificant, and for more than ~10 wires, `csr_matrix` is likely to be faster assuming that the matrix is sparse enough.

Reply:
Note that for anything involving default qubits Hamiltonian expectation computations and heuristics, I would ensure that @soranjh is involved, as he originally designed these heuristics using benchmarks based on quantum chemistry computations

Reply:
We did some initial benchmarks on random sparse matrices/vectors and @AmintorDusko did a more comprehensive benchmark on realistic qchem systems later in [this](https://github.com/PennyLaneAI/pennylane/pull/2561) and [this](https://github.com/PennyLaneAI/pennylane/pull/2630) PRs. Please note that the benchmarks are better to be on real qchem Hamiltonians not random sparse matrices/states.

Reply:
Just flying by, but, I would guess that for dense vectors it is more performant to keep it as numpy arrays and just converted the matrix. Both memory-wise and speed-wise. Some benchmarks, even if micro-benchmarks may be sensible.